src/lib/core/datetime/date-selection.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DateAdapter} from '@angular/material/core';
+
+export abstract class MatDateSelection<D> {
+  abstract add(date: D): void;
+  abstract clone(adaptor: DateAdapter<D>): MatDateSelection<D>;
+  abstract getFirstSelectedDate(): D|null;
+  abstract getLastSelectedDate(): D|null;
+  abstract isComplete(): boolean;
+  abstract isSame(adapter: DateAdapter<D>, other: MatDateSelection<D>): boolean;
+  abstract isValid(adapter: DateAdapter<D>): boolean;
+}
+
+export class MatSingleDateSelection<D> extends MatDateSelection<D> {
```

Comment:
We should add an `asDate` method to this so we can get a date to pass back to the user